diff --git a/Makefile b/Makefile
index adc9add..a4a4ce8 100644
--- a/Makefile
+++ b/Makefile
@@ -26,6 +26,18 @@ ifeq ($(ARCH),RV64)
 SAIL_DEFAULT_INST += riscv_insts_dext.sail riscv_insts_cdext.sail
 endif
 
+SAIL_DEFAULT_INST += riscv_insts_crypto.sail
+ifeq ($(ARCH),RV32)
+    SAIL_DEFAULT_INST += riscv_insts_crypto_rv32.sail
+else ifeq ($(ARCH),RV64)
+    SAIL_DEFAULT_INST += riscv_insts_crypto_rv64.sail
+else
+  $(error '$(ARCH)' is not a valid architecture, must be one of: RV32, RV64)
+endif
+#SAIL_DEFAULT_INST += riscv_insts_crypto_rvv_aes.sail
+#SAIL_DEFAULT_INST += riscv_insts_crypto_rvv_alu.sail
+#SAIL_DEFAULT_INST += riscv_insts_crypto_rvv_sha.sail
+
 SAIL_SEQ_INST  = $(SAIL_DEFAULT_INST) riscv_jalr_seq.sail
 SAIL_RMEM_INST = $(SAIL_DEFAULT_INST) riscv_jalr_rmem.sail riscv_insts_rmem.sail
 
@@ -63,6 +75,7 @@ SAIL_ARCH_SRCS = $(PRELUDE)
 SAIL_ARCH_SRCS += riscv_types_common.sail riscv_types_ext.sail riscv_types.sail
 SAIL_ARCH_SRCS += riscv_vmem_types.sail $(SAIL_REGS_SRCS) $(SAIL_SYS_SRCS) riscv_platform.sail
 SAIL_ARCH_SRCS += riscv_mem.sail $(SAIL_VM_SRCS)
+SAIL_ARCH_SRCS += riscv_types_crypto.sail
 SAIL_ARCH_RVFI_SRCS = $(PRELUDE) rvfi_dii.sail riscv_types_common.sail riscv_types_ext.sail riscv_types.sail riscv_vmem_types.sail $(SAIL_REGS_SRCS) $(SAIL_SYS_SRCS) riscv_platform.sail riscv_mem.sail $(SAIL_VM_SRCS)
 
 SAIL_STEP_SRCS = riscv_step_common.sail riscv_step_ext.sail riscv_decode_ext.sail riscv_fetch.sail riscv_step.sail
